addons/account/models/account_move.py

```diff
@@ -3895,28 +3895,28 @@ def check_full_reconcile(self):
         ):
 
             exchange_move_id = False
-            missing_exchange_difference = False
+            missing_exchange_difference = not self.env.context.get('no_exchange_difference')
             # Eventually create a journal entry to book the difference due to foreign currency's exchange rate that fluctuates
-            if to_balance and any([not float_is_zero(residual, precision_rounding=digits_rounding_precision) for aml, residual in to_balance.values()]):
-                if not self.env.context.get('no_exchange_difference'):
+            is_exchange_needed = missing_exchange_difference and any([not float_is_zero(residual, precision_rounding=digits_rounding_precision) for aml, residual in to_balance.values()])
+            if to_balance:
+                if is_exchange_needed:
                     exchange_move_vals = self.env['account.full.reconcile']._prepare_exchange_diff_move(
                         move_date=maxdate, company=amls[0].company_id)
                     if len(amls.mapped('partner_id')) == 1 and amls[0].partner_id:
                         exchange_move_vals['partner_id'] = amls[0].partner_id.id
                     exchange_move = self.env['account.move'].with_context(default_type='entry').create(exchange_move_vals)
                     part_reconcile = self.env['account.partial.reconcile']
-                    for aml_to_balance, total in to_balance.values():
-                        if total:
-                            rate_diff_amls, rate_diff_partial_rec = part_reconcile.create_exchange_rate_entry(aml_to_balance, exchange_move)
-                            amls += rate_diff_amls
-                            partial_rec_ids += rate_diff_partial_rec.ids
-                        else:
-                            aml_to_balance.reconcile()
+                for aml_to_balance, total in to_balance.values():
+                    if total and is_exchange_needed:
+                        rate_diff_amls, rate_diff_partial_rec = part_reconcile.create_exchange_rate_entry(aml_to_balance, exchange_move)
+                        amls += rate_diff_amls
+                        partial_rec_ids += rate_diff_partial_rec.ids
+                    else:
+                        aml_to_balance.reconcile()
```

Comment:
Could you give more info on what was going wrong exactly in this piece of code ? You're changing quite a bunch of things here, and it's honestly not obvious :p Is it really the minimal diff ?

Reply:
Sorry for the delay.
Here I'm trying to fix the case where `to_balance` looks like 
`{res.currency(2,): [account.move.line(23, 27), 0.0]}` 
In the given situation we will skip the whole block because of `any([not float_is_zero(residual, precision_rounding=digits_rounding_precision) for aml, residual in to_balance.values()])`, but actually there is an `amount_residual` in the 2 lines that should be reconciled

I changed the patch to have a minimal diff and added a test (although I'm not sure we want/need this anymore)